docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -294,254 +296,289 @@ TODO
+order, continuing with successive lanes in the `sortedLanes` array, that is, in decreasing priority
+order, and breaking the iteration when reaching `maxBytes` or `maxGas`.
+
+The mempool is locked during `ReapMaxBytesMaxGas`, so no transaction will be added or removed from
+the mempool during reaping.
+
+### Transaction dissemination
+
+For broadcasting transactions from multiple lanes, we have considered two possible approaches:
+1. Reserve $N$ p2p channels for use by the mempool. P2P channels have priorities that we can reuse
+   as lane priorities. There are a maximum of 256 P2P channels, thus limiting the number of lanes.
+2. Continue using the current P2P channel for disseminating transactions and implement logic within
+   the mempool to select the order of transactions to put in the channel. This option theoretically
+   allows for an unlimited number of lanes, constrained only by the nodes’ capacity to store the
+   lane data structures.
+
+We choose the second option for its flexibility, allowing us to start with a simple scheduling
+algorithm that can be refined over time. The desired algorithm must satisfy the properties
+"Priorities between classes" and "FIFO ordering per class". This requires supporting selection by
+_weight_, ensuring each lane gets a fraction of the P2P channel capacity proportional to its
+priority. Moreover, it should be _fair_, to prevent starvation of low-priority lanes. Given the
+extensive research on scheduling algorithms in operating systems and networking, we propose for the
+MVP implementing a variant of the [Weighted Round Robin (WRR)](wrr) algorithm, which meets these
+requirements and is straightforward to implement.
+
+Before modifying the dissemination logic, we need to refactor the current implementation and the
+`Mempool` interface to clearly separate the broadcast goroutine in the mempool reactor from
+`CListMempool` that includes the mempool data structures. `CListMempool` provides two methods used
+by the broadcast code, `TxsWaitChan() <-chan struct{}` and `TxsFront() *clist.CElement`, which are
+just wrappers around the methods `WaitChan` and `Front` of the `CList` implementation. In
+particular, `TxsFront` is leaking implementation details outside the `Mempool` interface.
+
+### Checking received transactions
+
+Transactions are transmitted without lane information because peers cannot be trusted to send the
+correct data. A node may take advantage of the network by sending lower-priority transactions before
+higher-priority ones. Although the receiving node could easily verify the priority of a transaction
+when it calls `CheckTx`, it cannot detect if a peer is sending transactions out of order over a
+single P2P channel. For the moment, we leave out of the MVP any mechanism to detect and possibly
+punish nodes for this kind of behaviour.
+
+## Alternative designs
+
+### One CList for all lanes
+
+We briefly considered sharing one CList for all lanes, changing the internal logic of CList to
```

Comment:
This is the priority mempool, in other words.

Reply:
I wouldn't say so because in the priority mempool every tx has a priority. The idea of the "one CList for all lanes" solution was something like adding a lane field to each entry and then other auxiliary data structures to support traversing the entries per lane.

Reply:
So it is not one clist for all lanes, but one clist and multiple clists, one per lane, using the same entries as the one big clist.

In other words, a linked-list that contains other abstract linked-lists inside it.